### DISCUSSION TITLE
How to bind `sort mtime` to a key?

### DISCUSSION THREAD

**Discussion by Kossak-esky** [leo-arch/clifm#302] [Q&A] [answered]
I tried adding `sort mtime:\M-x` to keybindings (F9), uncommenting existing shortcut first, but it doesn't work.

**Comment by leo-arch** [comment, marked as answer]
Hi @Kossak-esky!

For the time being, bindable actions are fixed, and there's no `sort mtime` action. However, it can be done by binding to a plugin. Follow these steps:

1. Create a new plugin named `sort_mtime.sh` under `~/.config/clifm/plugins` with this content:
```sh
#!/bin/sh
echo "sort mtime" > "$CLIFM_BUS"
exit 0
```
Do not forget to make it executable: `te sort_mtime.sh`.

2. Bind the new plugin to an action name: run `actions edit` and add this line: `plugin2=sort_mtime.sh`

3. Edit your keybindings and bind a key to the new plugin:
```
plugin2:\M-x
```
NOTE: There are four available slots for plugins (plugin[1-4]), and `plugin1` is already used by the `xclip.sh` plugin, which can be removed if you don't use it.

4. Restart **clifm**.

**Comment by tkossak** [comment]
Thank you very much! Is it possible to make this a toggle? When I press `M-x` it should sort by mtime, if I press it again it should sort by name?

**Reply by leo-arch** [reply]
Hi @tkossak. Yes, it is. You just need a way to keep track of the current sorting method, for example, using a temp file.

1. Create the `sort_mtime.sh` script:

```sh
#!/bin/sh

TEMPFILE="${TMPDIR:-/tmp}/clifm_sort_mtime"

if ! [ -f "$TEMPFILE" ]; then
	echo "sort mtime" > "$CLIFM_BUS"
	echo "1" > "$TEMPFILE"
	exit 0
fi

if [ "$(cat "$TEMPFILE")" = "1" ]; then
	echo "sort name" > "$CLIFM_BUS"
	echo "0" > "$TEMPFILE"
else
	echo "sort mtime" > "$CLIFM_BUS"
	echo "1" > "$TEMPFILE"
fi

exit 0
```
2. Make it executable: `te sort_mtime.sh`
3. Bind it to an action via `actions edit`: `plugin2=sort_mtime.sh`
4. Bind the action to a keybinding via `kb edit`: `plugin2:\M-x`.
5. Restart **clifm**.

I'm planning to export **clifm** status values via environment variables to make things like this easier in the future. So, instead of the temp file thing you would just need to check an environment variable (say **CLIFM_SORT_STYLE**).

**Reply by leo-arch** [reply]
With the [latest commit](https://github.com/leo-arch/clifm/commit/1cddad4ed85d9e73867089ebafc2ea5c736e0d88) status information is exported via environment variables, for example, the current sort style is exported as **CLIFM_SORT_STYLE**.

The `sort_mtime.sh` plugin can now be written as follows:

```sh
#!/bin/sh

if [ -z "$CLIFM_SORT_STYLE" ]; then
	echo "clifm: CLIFM_SORT_STYLE is not set" >&2
	exit 1
fi

if [ "$CLIFM_SORT_STYLE" = "mtime" ]; then
	echo "sort name" > "$CLIFM_BUS"
else
	echo "sort mtime" > "$CLIFM_BUS"
fi

exit 0
```

EDIT: For the complete list of exported values consult the [Environment section](https://github.com/leo-arch/clifm/wiki/Specifics#environment) in the Wiki.
